common/src/intrusive_circular_list.h

```diff
@@ -0,0 +1,662 @@
+     * @param list the list which the given items should be moved from
+     * @param move_first, move_last the range of items to move into this list
+     * @param move_count the number of items to move into this list
+     */
+    template <typename L>
+    intrusive_circular_list splice_replace(
+        iterator replace_first, iterator replace_last, const std::size_t replace_count,
+        L&& list, iterator move_first, iterator move_last, const std::size_t move_count) {
+
+        assert(empty() || contains(*replace_first));
+        assert(replace_last == end() || contains(*replace_last));
+        assert(replace_count > 0u);
+        assert(replace_count <= size());
+        assert(move_count > 0u);
+        assert(move_count <= list.size());
+
+        auto insert_position = replace_last;
+        auto result = remove(replace_first, replace_last, replace_count);
+        if (empty()) {
+            insert_position = end();
+        }
+
+        splice(insert_position, std::forward<L>(list), move_first, move_last, move_count);
+        return result;
+    }
+
+    /**
+     * Clears this list without deleting its items.
+     */
+    void release() {
+        m_head = nullptr;
+        m_size = 0u;
+    }
+
+    /**
+     * Clears this list and deletes all items.
+     */
+    void clear() {
+        if (!empty()) {
+            const auto get_link = GetLink();
+            auto cur = m_head;
+            do {
+                auto next = get_link(cur).next();
+                delete cur;
+                cur = next;
+            } while (cur != m_head);
+            m_size = 0u;
+        }
+    }
+private: // helpers
+    auto& get_link(T* item) const {
+        assert(item != nullptr);
+        return GetLink()(item);
+    }
+
+    T* get_previous(T* item) const {
+        assert(item != nullptr);
+        return get_link(item).previous();
+    }
+
+    T* get_next(T* item) const {
+        assert(item != nullptr);
+        return get_link(item).next();
+    }
+
+    void connect(T* previous, T* next) {
+        auto& previous_link = get_link(previous);
+        auto& next_link = get_link(next);
+
+        previous_link.set_next(next);
+        next_link.set_previous(previous);
+    }
+
+    /**
+     * Inserts the given items into this list before the given position. If this list is empty, then the items are
+     * inserted at the front of this list.
+     *
+     * @param position the position before which the items are inserted
+     * @param items the items to insert
+     * @param count the number of items to insert
+     */
+    void insert_before(iterator position, T* items, std::size_t count) {
+        if (empty()) {
+            m_head = items;
+            m_size = count;
+        } else {
+            auto previous = position == end() ? back() : get_previous(*position);
+            auto next = get_next(previous);
+
+            auto first = items;
+            auto last = get_previous(first);
```

Comment:
Mostly the auto usage in this PR looks good, but variables like these that are `T*` would be easier to read in git diffs if it was stated explicitly (and it's more concise).

Reply:
Yeah, I was debating this, but laziness prevailed. I'll use `T*` instead of `auto` where appropriate.